routes/routes.js

```diff
@@ -0,0 +1,122 @@
+const jwt = require('jsonwebtoken');
+
+const User = require('../models/userModel');
+const Note = require('../models/noteModel');
+
+const secret = 'ikea monkey';
+
+const authenticate = (req, res, next) => {
+  const token = req.get('authorization');
+  if (token) {
+    jwt.verify(token, secret, (err, decoded) => {
+      if (err) return res.status(422).json(err);
+      req.decoded = decoded;
+      next();
+    });
+  } else {
+    return res.status(403).json({
+      error: 'No valid token provided.'
+    });
+  }
+};
+
+const createUser = (req, res) => {
+  const { username, password } = req.body;
+  if (!username || !password) {
+    res
+      .status(422)
+      .json({ error: 'A valid username and password is required' });
+  } else {
+    const newUser = new User({ username, password });
+    newUser
+      .save()
+      .then(response => {
+        res.status(201).json(response);
+      })
+      .catch(err => {
+        res.status(500).json(err);
+      });
+  }
+};
+
+const login = (req, res) => {
+  const { username, password } = req.body;
+  User.findOne({ username: username.toLowerCase() }, (err, user) => {
+    if (err) {
+      res.status(403).json({ error: 'Invalid Username/Password' });
+      return;
+    }
+    if (user === null) {
+      res.status(422).json({ error: 'That User Does Not Exist' });
+      return;
+    }
+    user.verifyPassword(password, (notMatch, match) => {
```

Comment:
Here, our callback function, which we're passing to `bcrypt.compare` takes two arguments, `err` and `match` (refer to bcrypt docs: https://www.npmjs.com/package/bcrypt#to-check-a-password). We handle any errors that `bcrypt.compare` may encounter, and the second argument `match` is the result of comparing the plaintext password and the hashed password.

```JS
user.verifyPassword(password, (err, match) => {
  if (err) return res.status(500).json({ message: err.message });
  if (match) {
    const payload = { username: user.username };
    const token = jwt.sign(payload, secret);
    res.json({ token });
  } else {
    res.status(422).json({ error: "passwords don't match" });
  }
});
```